projects/scullyio/ng-lib/src/lib/transfer-state/transfer-state.service.ts

```diff
@@ -0,0 +1,133 @@
+import {HttpClient} from '@angular/common/http';
+import {Inject, Injectable} from '@angular/core';
+import {DOCUMENT} from '@angular/common';
+import {NavigationStart, Router} from '@angular/router';
+import {isScullyGenerated, isScullyRunning} from '../utils/isScully';
+import {Observable, of, Subject} from 'rxjs';
+import {catchError, filter, map, switchMap, tap} from 'rxjs/operators';
+
+const SCULLY_SCRIPT_ID = `scully-transfer-state`;
+const SCULLY_STATE_START = `___SCULLY_STATE_START___`;
+const SCULLY_STATE_END = `___SCULLY_STATE_END___`;
+
+@Injectable({
+  providedIn: 'root',
+})
+export class TransferStateService {
+  private script: HTMLScriptElement;
+  private state: {[key: string]: any} = {};
+  private fetching: Subject<any>;
+
+  constructor(
+    @Inject(DOCUMENT) private document: Document,
+    private router: Router,
+    private http: HttpClient
+  ) {
+    this.setupEnvForTransferState();
+    this.setupNavStartDataFetching();
+  }
+
+  private setupEnvForTransferState(): void {
+    if (isScullyRunning()) {
+      // In Scully puppeteer
+      this.script = this.document.createElement('script');
+      this.script.setAttribute('id', SCULLY_SCRIPT_ID);
+      this.script.setAttribute('type', `text/${SCULLY_SCRIPT_ID}`);
+      this.document.head.appendChild(this.script);
+    } else if (isScullyGenerated()) {
+      // On the client AFTER scully rendered it
+      this.script = this.document.getElementById(SCULLY_SCRIPT_ID) as HTMLScriptElement;
+      try {
+        this.state = JSON.parse(unescapeHtml(this.script.textContent));
+      } catch (e) {
+        this.state = {};
+      }
+    }
+  }
+
+  getState<T>(name: string): Observable<T> {
+    if (this.fetching) {
+      return this.fetching.pipe(map(() => this.state[name]));
+    } else {
+      return of(this.state[name]);
+    }
+  }
+
+  setState<T>(name: string, val: T): void {
+    this.state[name] = val;
+    if (isScullyRunning()) {
+      this.script.textContent = `${SCULLY_STATE_START}${escapeHtml(
+        JSON.stringify(this.state)
+      )}${SCULLY_STATE_END}`;
+    }
+  }
+
+  setupNavStartDataFetching() {
+    /**
+     * Each time the route changes, get the Scully state from the server-rendered page
+     */
+    if (!isScullyGenerated()) return;
+
+    this.router.events
+      .pipe(
+        filter(e => e instanceof NavigationStart),
+        tap(() => (this.fetching = new Subject<any>())),
+        switchMap((e: NavigationStart) => {
+          // Get the next route's page from the server
+          return this.http.get(e.url, {responseType: 'text'});
+        }),
+        catchError(err => {
+          console.warn('Failed transfering state from route', err);
+          return of('');
```

Comment:
@chrisguttandin Check it. Added a catchError on this transfer state fetching stream. Great call out. One failure here kills all future route change fetching as well.

Reply:
I think the `catchError` operator has to be "piped" to the `http.get` directly. Since RxJS can be tricky I created a little example to check if that is actually true. :-)

```typescript
interval(1000)
    .pipe(
        switchMap((_, index) => {
            if (index === 3) {
                return throwError(new Error());
            }

            return of(index);
        }),
        catchError(() => of(''))
    )
    .subscribe((x) => console.log(x))
```

The snippet above will log `1`, `2`, `''`.

```typescript
interval(1000)
    .pipe(
        switchMap((_, index) => {
            if (index === 3) {
                return throwError(new Error())
                    .pipe(catchError(() => of('')));
            }

            return of(index);
        })
    )
    .subscribe((x) => console.log(x))
```

This will log `1`, `2`, `''`, `4`, `5` ...

Reply:
To be clear, you are saying that I need to do the `http.get().pipe(catchError())`. Is that what you are saying you want to see?

Reply:
Mmm yes, I think that's necessary to keep the subscription to `this.router.events` alive. Otherwise it will be replaced by a subscription to of `of('')`.

Reply:
I see. Good call. Fixing now.

Reply:
@chrisguttandin check this latest change out. Let me know.

Reply:
👍 I think that works. What about unsubscribing from `this.router.events` in an `ngOnDestroy()` method? A `subscribe()` without an `unsubscribe()` always makes me a little nervous. :-)